### ISSUE TITLE
Can see through a ceiling from "dome"

### ISSUE THREAD

**Issue by slipher** [UnvanquishedAssets/map-parpax_src.dpkdir#19]
![unvanquished_2024-11-10_225228_000](https://github.com/user-attachments/assets/23c709e8-ba6f-4ead-97fd-82da3c719b8d)

From this location there is, geometrically speaking, a line of sight to many other rooms, as can be seen by turning on the material system. But the other ones are concealed by the skybox while this one isn't.

**Comment by slipher** [comment]
It's relying on vis apparently. With `r_novis` you can see a ton of stuff.

**Comment by sweet235** [comment]
> It's relying on vis

There are other maps doing this as well. One example is https://github.com/Masmblr/map-jota_src at `35 -1698 95 90 3` (this map is superseded by https://github.com/Masmblr/map-jota_src.dpkdir, which does not do this).

One could argue that the parts hidden by the sky are map deficiencies. After all, this must not happen in euclidean space. atcs, for example, erects two huge structures mainly for this reason. Nevertheless, I have wondered on occasion: is there a method to hide such rooms in some shader? Or can cleverly placed vis hints do it?

If we cannot find any method to make this room invisible, I see no other way than to either lower the height of the dome's ceiling, or to add nontransparent surfaces on top of the room we can look into.

**Comment by illwieckz** [comment]
Many maps suffer from this, uncreation has some of it (obvious), arachnid2 too (hard to spot). This is due to the skybox mechanism being as fast as it is naïve: the game first paints the sky, then paints anything else (even if the sky surface is in front of it). This is true since forever, tremulous/quake3 suffered from this. Of course it's a map bug since there can't be an hidden room behind the sky in real life, without an actual wall hidingit.

This feature by @VReaperV is meant to fix that:

- https://github.com/DaemonEngine/Daemon/pull/1299

See also:

- https://github.com/DaemonEngine/Daemon/issues/1249

**Comment by illwieckz** [comment]
> But the other ones are concealed by the skybox while this one isn't.

 > It's relying on vis apparently. With `r_novis` you can see a ton of stuff.

Indeed, there is no such thing as skybox concealing things, things not rendered are just things excluded by the vis.

**Comment by mjstahlberg** [comment]
Could it have been the case in the past that the skybox was hiding the geometry behind it? I was certainly up in the dome many times during testing and I never noticed this.

> One could argue that the parts hidden by the sky are map deficiencies. After all, this must not happen in euclidean space. atcs, for example, erects two huge structures mainly for this reason. 

fwiw, I considered there to be some (desert) landscape right around the dome and the hangar exit, with everything else being underground.

**Comment by sweet235** [comment]
> I considered there to be some (desert) landscape right around the dome and the hangar exit

That would probably be the easiest fix right now. Put a sand patch mesh there and call it a day.

**Comment by slipher** [comment]
I checked Unvanquished 0.49 and the bug was already there, so this was not caused by recent sky changes.

**Comment by illwieckz** [comment]
Yes, it's super old, there are old threads around about it:

> `2012-09-12` **illwieckz**: At one point you can see through the ceiling: https://postimg.cc/CnyLPnTM
> `2012-09-14` **Viech**: The first bug isn't my fault, it's based on the way the engine handles skyboxes. Whatever gets rendered from your perspective (which is not only what you can see but usually slightly more) can be seen through the skybox. I'd need to build an outside area that lays before the visible regions.
-- https://forum.tremulous.net/index.php?topic=16797.msg234786#msg234786

![map-parpax-beta3-bug](https://github.com/user-attachments/assets/304fcd07-aa79-4f6d-ab3f-bf7bb72a5713)

For archaeologist, it was linked there: https://github.com/Unvanquished/Unvanquished/issues/112
At the time I was saying the new renderer could not run on my computer. 🤭️

**Comment by illwieckz** [comment]
Funny to see that on the exact same hardware, the new renderer is now able to render the map, a bit slower on same resolution though:

![unvanquished_2024-11-12_002533_000](https://github.com/user-attachments/assets/726b7e84-05d6-4768-9925-374d771f4f5b)

With lower resolution it's more acceptable:

![unvanquished_2024-11-12_002621_000](https://github.com/user-attachments/assets/7d764a1b-2042-489d-bc0e-352f75590ac0)

The good news is that the elevator (the vertical lines on the right on screenshot from 2012) is now out of vis thanks to some vis optimizations having been done in the map since that time.

**Comment by VReaperV** [comment]
> This feature by [@VReaperV](https://github.com/VReaperV) is meant to fix that:
> 
> * [Generate skybrush material DaemonEngine/Daemon#1299](https://github.com/DaemonEngine/Daemon/pull/1299)
> 
> See also:
> 
> * [Material system: can see through walls on plat23 DaemonEngine/Daemon#1249](https://github.com/DaemonEngine/Daemon/issues/1249)

This only applies to material system. It's intended to basically simulate what vis is doing here, but these are, indeed, map bugs that were just hidden and entirely relied on how well vis was formed to conceal this geometry.
